src/aura/AuraAvatarUtils.sol

```diff
@@ -0,0 +1,163 @@
+        uint256 ethInUsd = fetchPriceFromClFeed(ETH_USD_FEED, CL_FEED_HEARTBEAT_ETH_USD);
+        // Divisor is 10^38 and uint256 max ~ 10^77 so this shouldn't overflow for normal amounts
+        usdcAmount_ = (_auraAmount * auraInEth * ethInUsd) / AURA_USD_TWAP_DIVISOR;
+    }
+
+    function getBalAmountInBpt(uint256 _balAmount) internal view returns (uint256 bptAmount_) {
+        uint256 bptPriceInBal = getBptPriceInBal();
+        bptAmount_ = (_balAmount * PRECISION) / bptPriceInBal;
+    }
+
+    function querySwapAuraForUsdc(uint256 _auraAmount) internal returns (uint256 usdcEarned_) {
+        IAsset[] memory assetArray = new IAsset[](3);
+        assetArray[0] = IAsset(address(AURA));
+        assetArray[1] = IAsset(address(WETH));
+        assetArray[2] = IAsset(address(USDC));
+
+        IBalancerVault.BatchSwapStep[] memory swaps = new IBalancerVault.BatchSwapStep[](2);
+        // AURA --> WETH
+        swaps[0] = IBalancerVault.BatchSwapStep({
+            poolId: AURA_WETH_POOL_ID,
+            assetInIndex: 0,
+            assetOutIndex: 1,
+            amount: _auraAmount,
+            userData: new bytes(0)
+        });
+        // WETH --> USDC
+        swaps[1] = IBalancerVault.BatchSwapStep({
+            poolId: USDC_WETH_POOL_ID,
+            assetInIndex: 1,
+            assetOutIndex: 2,
+            amount: 0, // 0 means all from last step
+            userData: new bytes(0)
+        });
+
+        IBalancerVault.FundManagement memory fundManagement = IBalancerVault.FundManagement({
+            sender: address(this),
+            fromInternalBalance: false,
+            recipient: payable(address(this)),
+            toInternalBalance: false
+        });
+
+        int256[] memory assetDeltas =
+            BALANCER_VAULT.queryBatchSwap(IBalancerVault.SwapKind.GIVEN_IN, swaps, assetArray, fundManagement);
+
+        usdcEarned_ = uint256(-assetDeltas[assetDeltas.length - 1]);
+    }
+
+    /// @notice Calculates the price of 80BAL-20WETH BPT in BAL using the BAL-ETH CL feed.
+    /// @return bptPriceInBal_ The price of 80BAL-20WETH BPT in BAL.
+    function getBptPriceInBal() internal view returns (uint256 bptPriceInBal_) {
+        uint256 ethPriceInBal = PRECISION_DOUBLE / fetchPriceFromClFeed(BAL_ETH_FEED, CL_FEED_HEARTBEAT_BAL);
+
+        uint256 invariant = IWeightedPool(address(BPT_80BAL_20WETH)).getInvariant();
+        uint256 totalSupply = BPT_80BAL_20WETH.totalSupply();
+
+        // p1: Price of BAL
+        // p2: Price of ETH
+        //        w1         w2
+        //  / p1 \     / p2 \
+        // |  --  | . |  --  |
+        //  \ w1 /     \ w2 /
+        // -------------------
+        //          p1
+        uint256 pricePerShare = LogExpMath.pow(
+            ethPriceInBal * W1_BPT_80BAL_20WETH / W2_BPT_80BAL_20WETH, W2_BPT_80BAL_20WETH
+        ) * PRECISION / W1_BPT_80BAL_20WETH;
+
+        bptPriceInBal_ = invariant * pricePerShare / totalSupply;
```

Comment:
How did you derive this formula? Do you have a reference for this?

Reply:
Sure
https://hackmd.io/@shuklaayush/BkAtKbCY9

I also added a test for this as a sanity check
https://github.com/Badger-Finance/badger-avatars/blob/3a44ea9d379a090e9479ab0efcb55bfd34f21448/test/aura/AuraAvatarTwoTokenTest.t.sol#L896-L903

Reply:
Sweet, this is gold. Derivation of formula to estimate price of BPT in terms of BAL is correct and leads to the most cost efficient way to calculate this (only one call of .pow()).

LG